### PR TITLE
Anvil Damage Implementation

### PR DESCRIPTION
**Pull request by nathanswanson**
## Introduction

When a player or an entity has an anvil drop on them no damage is applied to same entity.

### Relevant issues

[Issues 4437](https://github.com/pmmp/PocketMine-MP/issues/4437)

### API changes

No API Changes.

### Behavioural changes

- Anvils now create falling damage at a max of 40HP.  The damage is linked to the tick rate so odd behavior is limited.  
- EntityDamageEvent has been updated to contain "CAUSE_FALLING_ANVIL" 
- PlayerDeathEvent translation factory now has anvil death message.

_Nathan was squashed by an anvil_

## Backwards compatibility

Unlikely to have any issues.

## Follow-up

- Falling Block types could benefit to having an event to handle collisions for plugins.
- Sound when anvil finishes falling.
- Anvil functionality.
- Anvil piston behavior.

Requires translations:

| Name | Value in eng.ini |
| :--: | :---: |
| `death.attack.anvil` | `{%a} was squashed by an anvil` |

## Tests

Game Implementation so no PHPUnit test used. I tested it by dropping anvils at various heights on me and all mobs implemented.

### CODE REVIEW

File: src/entity/object/FallingBlock.php
Comment (thebigsmileXD) [inline src/entity/object/FallingBlock.php:113]: `getId()` and `===`

File: src/entity/object/FallingBlock.php
Comment (thebigsmileXD) [inline src/entity/object/FallingBlock.php:104]: `$this->fallTime += 1`

File: src/event/player/PlayerDeathEvent.php
Comment (thebigsmileXD) [inline src/event/player/PlayerDeathEvent.php:132]: `case`

**Review by inxomnyaa — CHANGES_REQUESTED** [review]
Styling

File: src/lang/KnownTranslationFactory.php
Comment (dktapps) [inline src/lang/KnownTranslationFactory.php:582]: This file is auto generated. Don't change it by hand.

File: src/lang/KnownTranslationKeys.php
Comment (dktapps) [inline src/lang/KnownTranslationKeys.php:131]: This file is auto generated. Don't change it by hand.

File: src/event/entity/EntityDamageEvent.php
Comment (dktapps) [inline src/event/entity/EntityDamageEvent.php:66]: FALLING_BLOCK would be better (in case more blocks like this are added in the future).

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:176]: Unnecessary formatting change.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:119]: I'm not very clear what the point of this check is. This is just going to prevent entities from receiving a second damage from falling anvil if their last damage was from an anvil, irrespective of which anvil caused it.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [reply src/entity/object/FallingBlock.php:104]: `fallTime++` is better.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:53]: This should have a type.

**Review by dktapps — CHANGES_REQUESTED** [review]

File: src/lang/KnownTranslationKeys.php
Comment (nathanswanson) [reply src/lang/KnownTranslationKeys.php:131]: How would I generate another translation key then?

File: src/event/entity/EntityDamageEvent.php
Comment (nathanswanson) [reply src/event/entity/EntityDamageEvent.php:66]: Yeah you bring up a good point I will change this.

File: src/lang/KnownTranslationKeys.php
Comment (alvin0319) [reply src/lang/KnownTranslationKeys.php:131]: https://github.com/pmmp/PocketMine-MP/blob/master/build/generate-known-translation-apis.php

File: src/entity/object/FallingBlock.php
Comment (nathanswanson) [reply src/entity/object/FallingBlock.php:119]: Yeah now that you mention it It is unexpected behavior. Its goal was to make sure an anvil only damages player once while falling otherwise it damages every tick. I'll fix this.

Comment (nathanswanson) [reply src/entity/object/FallingBlock.php:119]: I retrace my last statement it does damage entity RESPECTIVE to which anvil.

I tested this by dropping multiple anvils on me consecutively each one damaging every time.

File: src/lang/KnownTranslationKeys.php
Comment (dktapps) [reply src/lang/KnownTranslationKeys.php:131]: It has to be added to the English translation file first, and then the code generated with this script ^

**Comment by nathanswanson** [comment]
PHP CS Fixer is not working on my PHP Storm so for now I will try to fix manually.

File: src/event/entity/EntityDamageEvent.php
Comment (thebigsmileXD) [reply src/event/entity/EntityDamageEvent.php]: Dripstone i.e.

**Comment by nathanswanson** [comment]
Why does code style fail? on the actual php file its one line, on the log it shows + then -.

Also I made it so the autotranslation key is generated for the anvil death event. But now I get a generated code consistency error.

**Comment by dktapps** [comment]
> Why does code style fail? on the actual php file its one line, on the log it shows + then -.

You've left trailing whitespace on the line. Blank lines shouldn't have any indentation.

> Also I made it so the autotranslation key is generated for the anvil death event. But now I get a generated code consistency error.

You haven't added the string to the language file.

**Comment by dktapps** [comment]
You need to submit a pull request to the master branch of this repo to add the string to `eng.ini`: https://github.com/pmmp/Language

Once that's done, the `resources/locale` submodule should be updated and the translation APIs regenerated.

**Comment by nathanswanson** [comment]
I added the PR. This PR should be good to go after that translation is merged.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:115]: This would be better in `Anvil->tickFalling()`.

Comment (dktapps) [reply src/entity/object/FallingBlock.php:115]: On second look, this is in completely the wrong place. Anvil damage should only apply when the anvil entity actually turns back into a block.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [reply src/entity/object/FallingBlock.php:121]: Checking block identity is not reliable. It depends on the World's block cache, which might have been cleared since the damage occurred.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:118]: Braces must be used.

Comment (dktapps) [reply src/entity/object/FallingBlock.php:118]: This variable is also redundant. The above statement could be moved directly into the foreach.

File: src/entity/object/FallingBlock.php
Comment (thebigsmileXD) [reply src/entity/object/FallingBlock.php:115]: That is not the case for dripstone, it damages everything in it's path and i am pretty sure anvils also work this way.

Comment (nathanswanson) [reply src/entity/object/FallingBlock.php:115]: I will switch it but I feel like this is a case where the vanilla implementation is worse

Comment (thebigsmileXD) [reply src/entity/object/FallingBlock.php:115]: On a side note, standing on slabs does not prevent you from getting the damage

**Comment by nathanswanson** [comment]
Ideally I wanted to implement the damage code into the anvil class but having a falling block makes is a little more tricky. I put the code in the FallingBlock->onGround() method.

**Comment by nathanswanson** [comment]
After debugging workspace I FINALLY got my checkers to all work now.

**Comment by nathanswanson** [comment]
Strange, my coding checker didn't detect /** var */ as a problem and ill fix, regardless this PR cannot be merged until language/pmmp gets pulled so I recommend this be marked as on hold or whatever equivalent

**Comment by dktapps** [comment]
You can pull it yourself, if you want to.

```
cd resources/locale
git checkout master
git pull
cd ../..
git add resources/locale
```
and then commit as you would normally.

File: src/block/Anvil.php
Comment (dktapps) [inline src/block/Anvil.php:67]: This change should be removed (it's not relevant to the goal of the PR).

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:165]: I still think this should be delegated to the block, similarly to how `tickFalling()` is handled.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:170]: I'm not completely sure `EntityDamageByBlockEvent` suits here, since it's actually the entity doing the damage.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:114]: This change has no relevance to the PR objective.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:80]: This change has no relevance to the PR objective.

File: src/entity/object/FallingBlock.php
Comment (dktapps) [inline src/entity/object/FallingBlock.php:58]: This change has no relevance to the PR objective.

**Comment by nathanswanson** [comment]
Alright what about something like this then, fallable now has a method that is called when a falling entity hits the ground.

File: src/block/utils/Fallable.php
Comment (thebigsmileXD) [inline src/block/utils/Fallable.php:41]: I would not pass the bounding box here. I would instead replace all parameters with FallingBlock, since you can just get the parameters from there PLUS see which block hit the ground. This should probably also call an event of some sorts. PHPDoc comment too big.

**Review by inxomnyaa — CHANGES_REQUESTED** [review]

File: src/block/Anvil.php
Comment (thebigsmileXD) [inline src/block/Anvil.php:107]: I have been thinking about the damage calculation of this. It appears the minimum fall distance (from survival tests) to do damage is >1 block (~1.3 damages you, tested by standing on an enchanting table). So instead of the min() there should also be a check for this.

**Review by inxomnyaa — CHANGES_REQUESTED** [review]

File: src/block/Anvil.php
Comment (nathanswanson) [reply src/block/Anvil.php]: Added a check for making sure drop is >1 block

File: src/block/Anvil.php
Comment (thebigsmileXD) [inline src/block/Anvil.php:108]: Should the damage be casted to int? Or is passing floats valid / correct in the case of falling blocks?

**Comment by thebigsmileXD** [comment]
> Wearing a helmet reduces the damage by 1⁄4, but this costs durability on the helmet

Idk if there is a BlockSolidifyEvent of some sorts, it would be useful to invoke it with the falling block's block and the resulting block being the damaged block, so plugins can cancel the breaking.

Plus anvil destruction https://minecraft.fandom.com/wiki/Anvil#Becoming_damaged

**Comment by nathanswanson** [comment]
> > Wearing a helmet reduces the damage by 1⁄4, but this costs durability on the helmet
> 
> 
> 
> Idk if there is a BlockSolidifyEvent of some sorts, it would be useful to invoke it with the falling block's block and the resulting block being the damaged block, so plugins can cancel the breaking.
> 
> 
> 
> Plus anvil destruction https://minecraft.fandom.com/wiki/Anvil#Becoming_damaged 

While adding the helmet damage makes sense I feel like the rest of the listed implementations here  are unrelated to the PR.

File: src/block/Anvil.php
Comment (nathanswanson) [reply src/block/Anvil.php:108]: Does vanilla implement the damage as integers? I assumed they also do floating point damage.

**Comment by nathanswanson** [comment]
Note: Latest change exposes Living::damageItem()

File: src/block/Anvil.php
Comment (thebigsmileXD) [inline src/block/Anvil.php:120]: else flagForDespawn()?

Comment (nathanswanson) [reply src/block/Anvil.php:120]: flagForDespawn is controlled in Fallingblock.php already.

File: src/block/Anvil.php
Comment (thebigsmileXD) [inline src/block/Anvil.php:112]: What happened here, reverted strict check?

**Review by inxomnyaa — CHANGES_REQUESTED** [review]

File: src/block/Anvil.php
Comment (nathanswanson) [reply src/block/Anvil.php:112]: Because I got this PHPStan error.

`Error: Strict comparison using !== between pocketmine\item\Item and null will always evaluate to true.`

Comment (Rush2929) [reply src/block/Anvil.php:112]: The return value of `getHelmet()` is always Item.

Comment (nathanswanson) [reply src/block/Anvil.php]: Yeah thats why I decided to just remove the check.

File: src/block/Anvil.php
Comment (dktapps) [inline src/block/Anvil.php:110]: The damage is reduced **by** 1/4, not **to** 1/4 (at least, according to the wiki).

File: src/block/Anvil.php
Comment (dktapps) [inline src/block/Anvil.php:114]: I don't like this. At this point the damage has not yet been done and might be cancelled. This denies plugins the opportunity to prevent the damage to health and armour via EntityDamageEvent cancelling.

File: src/block/Anvil.php
Comment (nathanswanson) [reply src/block/Anvil.php:110]: Yeah I tested it you are correct.

**Comment by dktapps** [comment]
Closing due to lack of activity.

**Pull request closed without merging**
